src/coreclr/vm/threads.cpp

```diff
@@ -852,13 +798,111 @@ Thread* SetupThread()
         FastInterlockOr((ULONG *) &pThread->m_State, Thread::TS_TPWorkerThread);
     }
 
+    // Initialize the thread for the platform as the final step.
+    if (performPlatformInit)
+        pThread->InitPlatformContext();
+
 #ifdef FEATURE_EVENT_TRACE
     ETW::ThreadLog::FireThreadCreated(pThread);
 #endif // FEATURE_EVENT_TRACE
 
     return pThread;
 }
 
+Thread* SetupThread()
+{
+    WRAPPER_NO_CONTRACT;
+    return SetupThreadWorker(/* performPlatformInit */ true);
+}
+
+Thread* SetupMainThread()
+{
+    WRAPPER_NO_CONTRACT;
+
+#ifdef DEBUG
+    static bool isCalled = false;
+    _ASSERTE(!isCalled);
+    isCalled = true;
+#endif // DEBUG
+
+    // Initializing the platform cannot be done at this point since main thread
+    // initialization occurs prior to loading managed code. Since it is desirable
+    // to use managed code for some platform initialization we need to defer this
+    // initialization and instead require the main thread to call InitPlatformContext()
+    // at an appropriate time.
+    return SetupThreadWorker(/* performPlatformInit */ false);
+}
+
+static Thread* SetupThreadNoThrowWorker(HRESULT *pHR, bool performPlatformInit)
+{
+    CONTRACTL {
+        NOTHROW;
+        if (GetThreadNULLOk()) {GC_TRIGGERS;} else {DISABLED(GC_NOTRIGGER);}
+    }
+    CONTRACTL_END;
+
+    HRESULT hr = S_OK;
+
+    Thread *pThread = GetThreadNULLOk();
+    if (pThread != NULL)
+    {
+        return pThread;
+    }
+
+    EX_TRY
+    {
+        pThread = SetupThreadWorker(performPlatformInit);
+    }
+    EX_CATCH
+    {
+        // We failed SetupThread.  GET_EXCEPTION() may depend on Thread object.
+        if (__pException == NULL)
+        {
+            hr = E_OUTOFMEMORY;
+        }
+        else
+        {
+        hr = GET_EXCEPTION()->GetHR();
+    }
+    }
+    EX_END_CATCH(SwallowAllExceptions);
+
+    if (pHR)
+    {
+        *pHR = hr;
+    }
+
+    return pThread;
+}
+
+//-------------------------------------------------------------------------
+// Public function: SetupThreadNoThrow()
+// Creates Thread for current thread if not previously created.
+// Returns NULL for failure (usually due to out-of-memory.)
+//-------------------------------------------------------------------------
+Thread* SetupThreadNoThrow(HRESULT *pHR)
+{
+    WRAPPER_NO_CONTRACT;
+    return SetupThreadNoThrowWorker(pHR, /* performPlatformInit */ true);
```

Comment:
Are all places that `SetupThreadNoThrow` called on threads created by the runtime? I think that I can see number of places where this is called for external threads.

I am wondering whether the logic to identify threads created by runtime (vs. external threads) should be more explicit. We should only needs this on these threads:
- Main thread
- Finalizer thread
- new Thread
- Threadpool threads

Is that right?

Reply:
>Are all places that `SetupThreadNoThrow` called on threads created by the runtime? I think that I can see number of places where this is called for external threads.

There are several outstanding but those all appear to be on COM scenarios - which is long standing behavior. I am more than willing to handle that in a subsequent PR as there are other things I would like to clean up.

>I am wondering whether the logic to identify threads created by runtime (vs. external threads) should be more explicit.

That list is what I have in mind too. I can narrow that down more if desired but then we are getting into some COM related behavioral changes that I would like to avoid at present. I think for non-Windows platforms this is correct and for the Windows platform this is correct in spirit.

Reply:
If it'll be made more explicit, I think the tiering and multi-core JIT background threads may also need this, as they may run some managed code while jitting.

Reply:
Why not treat tiereing background thread the same way as external thread that wandered into the runtime?

Reply:
I figure external threads are expected to provide their own pool, whereas these background threads are created by the runtime.

Reply:
I see, in that case it would probably be ok to not have a pool for the background jitting threads

Reply:
Is this a priority to fix in this PR or can it be deferred?

Reply:
I think that the whole PR can be signficantly simplified with this plan.
- The runtime created threads can be handled by adding the callback to `Thread.StartHelper.RunWorker`. No unmanaged handshake required.
- The main thread can be handled by adding the callback to the same place as `StartupHookProvider.RunStartupHooks`
- The finalizer thread is the only one that needs a bit more complex handshake to delay the start

Reply:
> The runtime created threads can be handled by adding the callback to Thread.StartHelper.RunWorker. No unmanaged handshake required.

Done.

> The main thread can be handled by adding the callback to the same place as StartupHookProvider.RunStartupHooks

That is possible but we still need to clean up the main thread from native. Given that I am inclined to keep it all in native for both runtimes since it is logically symmetrical now. All managed Threads are done in managed and the Main and Finalizer thread are both handled on the native side.

Reply:
It is fine to call this from native. I was more concerned about moving this closer to the point where we call Main method, and not during EEStartup. 

We have hosting entrypoint that initializes the runtime, without setting up the program entrypoint or calling the Main method (ie it just calls EEStartup, and not the rest). I do not think that it makes sense for this hosting entrypoint to setup the autorelease pool. I think it would be more appropropriate to setup the autorelease pool only once we are actually about to call the Main method. 

It matches what we do for COM: EEStartup does not initialize the COM on the thread. COM is only initialized only once we are about to call the Main method.